[transmission] Add new keybindings

arifer612 opened the pull request:
This commit adds 3 sets of keybindings and organises the keybindings so that it can be read easier.

1. Marking and labelling
   =SPC m t m= marks a torrent and =SPC m s l= sets the labels for marked torrents. The old command bound to =SPC m s l=, =transmission-set-ratio=, is bound to =SPC m s r= instead.
2. Queueing
   Keybindings under a new prefix are added to ~transmission-mode~ to manipulate how the torrents are queued. A keybinding to sort logically by columns is also added.
3. Turtle mode
   A new keybinding while in =transmission-mode= toggles the =transmission-turtle-mode= minor mode. When it is active, a new set of 3 keybindings will be active to change the settings of the minor mode. The bindings used are uppercase letters since there are already defined bindings using the lowercases.

Although there is a keybinding for =transmission-remove= in =transmission-mode=, a new keybinding for =transmission-delete= is also added. This is done since =transmission-remove= merely removes the torrent from Transmission but does not delete the file(s).

The changes are added to =CHANGELOG.develop=

lebensterben commented:
I suggest to use bindings defines here https://github.com/emacs-evil/evil-collection/blob/master/modes/transmission/evil-collection-transmission.el

arifer612 commented:
The bindings that I have added are not in evil-collection either. Things like queue ordering, labelling, and turtle mode, I added them because I see them as fundamental API used in managing active torrents in Transmission (except perhaps =transmission-delete= but I added that just for completion).

lebensterben commented:
I'm not saying you shouldn't add anything not included in evil-collections already.

I suggest you to use the same bindings as in evil-collections, if a function is defined both in evil-collections and in your PR.

And also it's best to avoid conflicts with evil-collections.

lebensterben commented on CHANGELOG.develop:
```diff
@@ -3615,6 +3619,21 @@ files (thanks to Daniel Nicolai)
   (thanks to Aaron Jensen)
+**** Transmission
+- Key bindings (thanks to Arif Er):
+  - ~SPC m d~ is to remove and delete torrent at point.
```
this for example is strictly worse than evil-collections'.

removal is a destructive operation and is better bounded to D (S-d)

arifer612 commented:
I see, I'll make the appropriate corrections then!
So for the example you pointed out, the remove-and-delete should be bound to `SPC m D` instead, is that right?

arifer612 commented:
I've changed the bindings for `transmission-delete` and `transmission-remove` to use the uppercase letters since they are both destructive commands. The other prior bindings differ from the bindings used by evil-collection but they can be intuitively understood with the current bindings and are not destructive.

lebensterben commented on layers/+tools/transmission/packages.el:
```diff
@@ -29,50 +29,94 @@
+              "S"  'tabulated-list-sort
+
+              ;; add
               "aa" 'transmission-trackers-add
```
"aa" and "at" are confusing to me..

Why not `aa` for `transmission-add` and `at` for `transmission-trackers-add`?

arifer612 replied:
To be honest, I agree with you here. These bindings were already present from before and I believe the author meant for it to match the binding for `transmission-tracker-add` in `transmission-info-mode`. It would make more sense to change them right?

lebensterben commented on layers/+tools/transmission/packages.el:
```diff
@@ -29,50 +29,94 @@
+              ;; goto
+              "gi" 'transmission-info
+              "gp" 'transmission-peers
+              "gt" 'transmission-files
```
what does `t` in `gt` mean?

arifer612 replied:
This binding was also brought over as it was from before so I do not know the reason for it. I can change it to `gf` for `[g]o to [f]iles`

lebensterben commented on layers/+tools/transmission/packages.el:
```diff
@@ -29,50 +29,94 @@
+              "gt" 'transmission-files
+
+              ;; set
+              "sb" 'transmission-set-bandwidth-priority
```
should this be "sp" as for set priority?
bandwidth cannot be set.

arifer612 replied:
This was also brought over as it was from before. I failed to spot that it does not match with the binding in `transmission-files-mode` where it is bound to `sp`. This should be changed.

lebensterben commented on layers/+tools/transmission/packages.el:
```diff
@@ -29,50 +29,94 @@
+              "ts" 'transmission-toggle
+              "tt" 'transmission-turtle-mode
+
+              ;; Queue
```
How about
J/K move down/move up
H/L move to top/move to bottom

arifer612 replied:
I thought of doing this but I was unsure if it would okay for me to use the movement keys here. If we're using HJKL should it be nested under a prefix like "SPC m Q H" "SPC m Q J" ... or would "SPC m H" "SPC m J" ... be enough?

lebensterben replied:
well, putting them under `SPC m` is still very cumbersome.
How about define them as evil normal state keys?
`(evil-define-key 'normal transmission-mode-map ..)`

lebensterben replied:
Also, please move this entire chunk except `(spacemacs/set-leader-keys "att" 'transmission)` to config, so that they're executed only when `transmission` is loaded.

arifer612 replied:
I will make the changes soon.

arifer612 replied:
I've moved everything but the binding for "att" to `:config` and changed the queue manipulation keys to use HJKL. Please check if they're in order.

arifer612 commented:
Made the necessary corrections as recommended!

arifer612 commented:
In the latest squash, I've moved everything but the binding for "att" to :config and changed the queue manipulation keys to use HJKL. Please check if they're in order.

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
-| ~SPC m s b~ | Set bandwidth priority of torrent(s) at point or in region.   |
+| Key binding | Description                                                       |
+|-------------+-------------------------------------------------------------------|
+| ~H~         | Move torrent to the top of the queue.                             |
```
You can group J/K and H/L so it's clearer

arifer612 replied:
This is an easy task, I'll get this done up!

arifer612 replied:
This has been settled!

lebensterben commented on layers/+tools/transmission/packages.el:
```diff
@@ -27,56 +27,99 @@
+        "S"  'tabulated-list-sort
+
+        ;; add
+        "at" 'transmission-trackers-add
```
since "A" is not used, how about
"a" transmission-add
"A" transmission-trakers-add

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m m~   | Move torrent at point or in region to a new location.             |
+| ~SPC m q~   | Quit and bury the buffer.                                         |
+| ~SPC m v~   | Verify torrent at point or in region.                             |
+| ~SPC m D~   | Prompt to remove and delete torrent at poit or torrens in region. |
```
description of remove and delete needs to be improved
the difference between them is `transmission-remove` only remove the torrent by default, while `transmission-delete` in addition remove local data. (`transmission-remove` prompts whether to remove local data)

See https://github.com/transmission/transmission/blob/8566df069899ce8923463cadeb0ff66d4544991a/extras/rpc-spec.txt#L439

arifer612 replied:
I'll work on the description for these 2. The reference you provided is perfect! I never thought of checking the source for Transmission.

The reason why I thought of changing the binding to "R" and "D" respectively was because they're both destructive executions (removing data from the client/disk). Both `transmission-remove` and `transmission-delete` prompt the user for confirmation of the removal (and deletion) before carrying out the process.

arifer612 replied:
I've made the descriptions a little clearer. I've left the bindings as "R" and "D" for the purpose of having `transmission-move` and `transmisison-mark` not clash with each other.

    | `SPC m D`    | Delete from disk and remove torrent(s) at point or in region.     |
    | `SPC m R`    | Remove torrent(s) at point or in region.                          |

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m D~   | Prompt to remove and delete torrent at poit or torrens in region. |
+| ~SPC m R~   | Prompt to remove torrent at point or torrents in region.          |
+| ~SPC m S~   | Sort torrents according to the column at point.                   |
+| ~SPC m a a~ | Add torrent by filename, URL, magnet link, or info hash.          |
```
above each prefix you can add a heading, for example
```txt
| ~SPC m S~   | Sort torrents according to the column at point.                   |
|-------------+-------------------------------------------------------------------|
| Add         |                                                                   |
|-------------+-------------------------------------------------------------------|
| ~SPC m a a~ | Add torrent by filename, URL, magnet link, or info hash.          |
```

arifer612 replied:
Adding headers is an easy task as well. It'll make reading the README much smoother, got it.

I'll remove `transmission-trackers-add` from `transmission-mode-map` and just leave it in `transmission-info-mode-map` where trackers are listed explicitly.

arifer612 replied:
After checking on the documentation for `transmission-trackers-add`, I believe that it should remain in `transmission-mode-map` because it should be used to add trackers to the selected torrent, or multiple marked torrents even in `transmission-mode`. For that reason, we should leave `SPC m a` as a prefix for `SPC m a a` and `SPC m a t`.

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m S~   | Sort torrents according to the column at point.                   |
+| ~SPC m a a~ | Add torrent by filename, URL, magnet link, or info hash.          |
+| ~SPC m a t~ | Add announce URLs to torrent or torrents.                         |
+| ~SPC m g i~ | Open a =transmission-info-mode= buffer for torrent at point.      |
```
for the group, `SPC m <i | f | p>`
the default bindings are
- `transmission-info`:  `i`
- `transmission-files`: `<RET>` or `!`
- `transmission-peers`: `e`

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m s l~ | Set label(s) for marked torrent(s) or torrent at point.           |
+| ~SPC m s r~ | Set global seed ratio limit.                                      |
+| ~SPC m s u~ | Set global upload speed limit in kB/s.                            |
+| ~SPC m t m~ | Toggle mark on torrent(s) at point or in region.                  |
```
I think we can make toggle mark `SPC m m`, whose default binding is `m`

arifer612 replied:
I had thought of having it in `SPC m t m` because it was a toggle. But if having it out in `SPC m m ` then I believe adding `SPC m x` for `transmission-toggle` (toggle between activating and deactivating a torrent process) should be added too.

arifer612 replied:
I have just remembered that ~SPC m m~ to toggle marks would clash with ~transmission-move~.

arifer612 replied:
Tying with the comment about fixing the keybindings for ~transmission-remote~ and ~transmission-delete~, we can just have
"R": ~transmission-remove~ for "[R]emove"
"D": ~transmission-delete~ for "[D]elete"
"r": ~transmission-move~ for "[r]ename"
"m": ~transmission-toggle-mark~ for "[m]ark"

arifer612 replied:
Made the changes as mentioned in the comment above this.

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m s u~ | Set global upload speed limit in kB/s.                            |
+| ~SPC m t m~ | Toggle mark on torrent(s) at point or in region.                  |
+| ~SPC m t s~ | Toggle torrent between started and stopped.                       |
+| ~SPC m t t~ | Toggle =transmission-turtle-mode=.                                |
```
Needs to add key bindings for turtle-mode as well:
"sD" 'transmission-turtle-set-days
"sS" 'transmission-turtle-set-speeds
"sT" 'transmission-turtle-set-times

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
 
 | Key binding | Description                                                   |
 |-------------+---------------------------------------------------------------|
 | ~SPC m c~   | Copy magnet link of current torrent.                          |
```
we should add an evil normal state binding to copy the magnet link.
it should be "y".

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
-| ~SPC m s l~ | Set seed ratio limit of torrent(s) at point.                  |
 | ~SPC m m~   | Move torrent at point or in region to a new location.         |
-| ~SPC m a a~ | Add announce URLs to torrent or torrents.                     |
 | ~SPC m T~   | Remove trackers from torrent at point by ID or announce URL.  |
```
remove tracker can just be `SPC m D`
(As mentioned in previous comment I recommend to have `SPC m a` for adding trackers in `transmission-info-mode`, then these two bindings pair very well.)

arifer612 replied:
`SPC m T` is used here to match the binding used in evil-collections, `T` for "Remove (T)racker". Using `SPC m D` here might cause some confusion as it is the same binding for `transmission-delete` in `transmission-mode-map`

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
 | ~SPC m T~   | Remove trackers from torrent at point by ID or announce URL.  |
-| ~SPC m s u~ | Set upload limit of torrent(s) at point in kB/s.              |
+| ~SPC m a t~ | Add announce URLs to torrent or torrents.                     |
+| ~SPC m g p~ | Open a =transmission-peers-mode= buffer for torrent at point. |
```
we might as well add `SPC m a f` here

arifer612 replied:
Sure thing, this will then match up with the mode-networkings that were provided in the default bindings.

arifer612 replied:
Added `SPC m g f` here!

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m g p~ | Open a =transmission-peers-mode= buffer for torrent at point.     |
+| ~SPC m s p~ | Set bandwidth priority of torrent(s) at point or in region.       |
+| ~SPC m s d~ | Set global download speed limit in kB/s.                          |
+| ~SPC m s l~ | Set label(s) for marked torrent(s) or torrent at point.           |
```
the default binding for setting ratio limit and setting labels are "l" and "L" respectively.

We should just use them.

arifer612 replied:
I am in favour of "r" and "l" for "(r)atio limit" and "(l)abel" because I have been using a function I made bound to "L" to set "(L)abels by regex", but this then would be different from both the default bindings and evil-collections.

They are consistent in all mode maps, only the README file was not updated properly as I mentioned earlier above.

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
+| ~SPC m a t~ | Add announce URLs to torrent or torrents.                     |
+| ~SPC m g p~ | Open a =transmission-peers-mode= buffer for torrent at point. |
+| ~SPC m s d~ | Set download limit of torrent(s) at point in kB/s.            |
+| ~SPC m s l~ | Set seed ratio limit of torrent(s) at point.                  |
```
we should also add `SPC m s L` for adding labels.

arifer612 replied:
Oh, it has already been added but I forgot to correct this table! What a blunder on my end!

arifer612 replied:
Corrected the README file

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
 | Key binding | Description                                                   |
 |-------------+---------------------------------------------------------------|
-| ~SPC m g f~ | Visit the file at point with =find-file-read-only=.           |
 | ~SPC m r~   | Run a command on the file at point.                           |
```
it'd better be "!", which parallels with `SPC !` which runs shell command

arifer612 replied:
"!" is already provided in evil-collection and is the default binding provided by the package as you mentioned. I will change this to `SPC m X` for "E(X)ECUTE in the major mode".

arifer612 replied:
Changed the key binding to `SPC m X` as proposed

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
 | ~SPC m g i~ | Open a =transmission-info-mode= buffer for torrent at point.  |
-| ~SPC m m~   | Move torrent at point or in region to a new location.         |
+| ~SPC m g p~ | Open a =transmission-peers-mode= buffer for torrent at point. |
+| ~SPC m m m~ | Move torrent at point or in region to a new location.         |
```
we should remove `SPC m m m`. normally user won't try to remove a torrent in files view.
also by default there's no such binding in transmission-files-mode-map

arifer612 replied:
I believe that the original author had meant for marked files of a multi-file torrent to be moved to a different location with this binding. `transmission-move` does not serve this purpose so I will remove it.

arifer612 replied:
Removed this binding.

lebensterben commented on layers/+tools/transmission/README.org:
```diff
@@ -63,52 +64,68 @@ Add =(transmission :variables transmission-auto-refresh-all t)= to
-| ~SPC m m~   | Move torrent at point or in region to a new location.         |
+| ~SPC m g p~ | Open a =transmission-peers-mode= buffer for torrent at point. |
+| ~SPC m m m~ | Move torrent at point or in region to a new location.         |
 | ~SPC m m u~ | Mark file(s) at point or in region as unwanted.               |
```
for wanted and unwanted, we can use `SPC m t w/u` instead.
this frees up `SPC m m` for `transmission-toggle-mark`.

arifer612 replied:
That's a good idea! This is an easy change to make.

arifer612 replied:
Changed the prefixes for these 2 bindings and `SPC m m` is reserved for `transmission-toggle-mark` to maintain consistency with the other modes.

lebensterben commented:
Looks pretty good. See my new comments

arifer612 commented:
I've made several changes. Only those that were not marked resolved are left as they are, with my reasons for so in the thread. I'd like to hear more about your thoughts about my opinions for not changing them.

I will not be squashing all the commits at once anymore until everything is finalised. I realised that it gets messy if I just squash everything every time I make a change for review.

lebensterben commented:
@arifer612 
Don't worry about squashing. I can squash and merge.

arifer612 commented:
Are there any comments regarding the latest version of the change?

lebensterben commented:
@arifer612 
Sorry for the delay. It's ready to merge now.
Do you want to update the commit message?

lebensterben commented:
@arifer612 
Terrific. That commit message is well-written.

I've just merged this PR. Thanks for your contribution.

arifer612 commented:
Thank you for the kind words! I really appreciate it!
